### ISSUE TITLE
panic: 64bit unaligned in arm32

### ISSUE THREAD

**Issue by nekohasekai** [etcd-io/bbolt#577]
Version: f3bb36466ac65b51d831a6eabcd78dde5f57feb7

```log
fatal error: sync: unlock of unlocked mutex

goroutine 11 [running]:
sync.fatal({0xb6e53f9e, 0x1e})
        $GOROOT/src/runtime/panic.go:1061 +0x24
sync.(*Mutex).unlockSlow(0x8e0a7034, 0xffffffff)
        $GOROOT/src/sync/mutex.go:229 +0x38
sync.(*Mutex).Unlock(0x8e0a7034)
        $GOROOT/src/sync/mutex.go:223 +0x44
go.etcd.io/bbolt.(*Tx).close(0x8e145cc0)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/tx.go:314 +0x1e4
go.etcd.io/bbolt.(*Tx).rollback(0x8e145cc0)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/tx.go:299 +0x13c
go.etcd.io/bbolt.(*DB).Update.func1()
        pkg/mod/go.etcd.io/bbolt@v0.0.0/db.go:858 +0x28
panic({0xb7d08ea8, 0xb7e7c630})
        $GOROOT/src/runtime/panic.go:914 +0x26c
runtime/internal/atomic.panicUnaligned()
        $GOROOT/src/runtime/internal/atomic/unaligned.go:8 +0x2c
runtime/internal/atomic.Xadd64(0x8e0a6fa4, 0x1)
        $GOROOT/src/runtime/internal/atomic/atomic_arm.s:258 +0x14
go.etcd.io/bbolt.(*TxStats).IncPageCount(...)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/tx.go:683
go.etcd.io/bbolt.(*TxStats).add(0x8e0a6fa4, 0x8e145cf0)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/tx.go:642 +0xec
go.etcd.io/bbolt.(*Tx).close(0x8e145cc0)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/tx.go:322 +0x25c
go.etcd.io/bbolt.(*Tx).Commit(0x8e145cc0)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/tx.go:230 +0x450
go.etcd.io/bbolt.(*DB).Update(0x8e0a6f20, 0x8e04cf90)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/db.go:873 +0x100
go.etcd.io/bbolt.(*batch).run(0x8e150e20)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/db.go:981 +0x110
sync.(*Once).doSlow(0x8e150e28, 0x8e04cfe4)
        $GOROOT/src/sync/once.go:74 +0xc0
sync.(*Once).Do(0x8e150e28, 0x8e04cfe4)
        $GOROOT/src/sync/once.go:65 +0x40
go.etcd.io/bbolt.(*batch).trigger(...)
        pkg/mod/go.etcd.io/bbolt@v0.0.0/db.go:963
created by time.goFunc
        $GOROOT/src/time/sleep.go:176 +0x30
```

Related code: https://github.com/etcd-io/bbolt/blob/ef065626c30a379c5b336c96a8175abb19b78425/tx.go#L26-L43

**Comment by ahrtr** [comment]
It should a known issue of golang. FYI. https://pkg.go.dev/sync/atomic#pkg-note-BUG

**Comment by ahrtr** [comment]
@nekohasekai the arch is arm32, what's the OS?

**Comment by nekohasekai** [comment]
@ahrtr Nexus 7 2013 Android 7.1.2

**Comment by ahrtr** [comment]
> @ahrtr Nexus 7 2013 Android 7.1.2

thx for the feedback. It seems a known issue.

```
On non-Linux ARM, the 64-bit functions use instructions unavailable before the ARMv6k core.
```

**Comment by nekohasekai** [comment]
Do you have any plans to make it 64bit aligned? This problem does not exist in the stable version of bbolt, and there are a large number of users on these 32-bit devices.

**Comment by ahrtr** [comment]
> Do you have any plans to make it 64bit aligned? This problem does not exist in the stable version of bbolt, and there are a large number of users on these 32-bit devices.

- Can you reproduce this issue in [v1.3.7](https://github.com/etcd-io/bbolt/releases/tag/v1.3.7)?
- I don't have such device at hand. Would you mind to find out starting from which commit on the master branch you see this issue?

**Comment by nekohasekai** [comment]
It does not exist in 1.3.7.

Here is a simple way to reproduce this issue on Linux amd64 using `qemu-user-static`:

Code:

```go
// cmd/bug/main.go

package main

import (
	"os"

	"go.etcd.io/bbolt"
)

func main() {
	db, _ := bbolt.Open("test.db", 0600, &bbolt.Options{})
	defer db.Close()
	defer os.RemoveAll("test.db")
	db.Batch(func(tx *bbolt.Tx) error {
		tx.CreateBucketIfNotExists([]byte("test"))
		return nil
	})
}
```

```bash
GOOS=linux GOARCH=arm go build ./cmd/bug
qemu-arm ./bug
````

**Comment by ahrtr** [comment]
It seems that it's related to https://github.com/etcd-io/bbolt/pull/373

**Comment by nekohasekai** [comment]
It is introduced in https://github.com/etcd-io/bbolt/commit/8ebdaf8d2ca98e26cc81d5ea33a1a86dd7cffc5f

**Comment by ahrtr** [comment]
Confirmed that it has nothing to do with https://github.com/etcd-io/bbolt/pull/516. Instead, it's caused by https://github.com/etcd-io/bbolt/pull/373.

After I partially reverted the change in https://github.com/etcd-io/bbolt/pull/373 as below, then couldn't reproduce this issue anymore.
```
$ git diff
diff --git a/tx.go b/tx.go
index 8b86030..905960a 100644
--- a/tx.go
+++ b/tx.go
@@ -639,18 +639,18 @@ type TxStats struct {
 }
 
 func (s *TxStats) add(other *TxStats) {
-       s.IncPageCount(other.GetPageCount())
-       s.IncPageAlloc(other.GetPageAlloc())
-       s.IncCursorCount(other.GetCursorCount())
-       s.IncNodeCount(other.GetNodeCount())
-       s.IncNodeDeref(other.GetNodeDeref())
-       s.IncRebalance(other.GetRebalance())
-       s.IncRebalanceTime(other.GetRebalanceTime())
-       s.IncSplit(other.GetSplit())
-       s.IncSpill(other.GetSpill())
-       s.IncSpillTime(other.GetSpillTime())
-       s.IncWrite(other.GetWrite())
-       s.IncWriteTime(other.GetWriteTime())
+       s.PageCount += other.PageCount
+       s.PageAlloc += other.PageAlloc
+       s.CursorCount += other.CursorCount
+       s.NodeCount += other.NodeCount
+       s.NodeDeref += other.NodeDeref
+       s.Rebalance += other.Rebalance
+       s.RebalanceTime += other.RebalanceTime
+       s.Split += other.Split
+       s.Spill += other.Spill
+       s.SpillTime += other.SpillTime
+       s.Write += other.Write
+       s.WriteTime += other.WriteTime
 }
```

**Comment by nekohasekai** [comment]
Confirm fixed.

**Comment by ahrtr** [comment]
> It is introduced in [8ebdaf8](https://github.com/etcd-io/bbolt/commit/8ebdaf8d2ca98e26cc81d5ea33a1a86dd7cffc5f)

Eventually I understood the reason why the issue couldn't be reproduced even it still contains the change in https://github.com/etcd-io/bbolt/pull/373. Because one field "`filesz int`" was removed in the commit [8ebdaf8](https://github.com/etcd-io/bbolt/commit/8ebdaf8d2ca98e26cc81d5ea33a1a86dd7cffc5f), and somehow caused the first field [PageCount](https://github.com/etcd-io/bbolt/blob/233156a53aebc333337d8dc7b350271c9611591f/tx.go#L601) in  isn't 64bit aligned anymore. 

Adding "`filesz int`" back can fix this issue, but we should think about a more reliable solution.

**Comment by nekohasekai** [comment]
You can either migrate atomic calls to int64 variables to atomic.Int64 structs (requires a higher version of Golang, in older versions you can use the inefficient atomic.Value), or change the embedding method of the entire structure from struct to pointer.

**Comment by ahrtr** [comment]
Using `atomic.Int64` can resolve this issue, but it will also makes the `TxStats` uncopyable (Note `atomic.Int64` has a `_ noCopy` field). It also requires updating some other public methods, e.g. [(*TxStats) Sub](https://github.com/etcd-io/bbolt/blob/233156a53aebc333337d8dc7b350271c9611591f/tx.go#L659). It may have impact on user applications as well.

The https://github.com/etcd-io/bbolt/pull/582 seems to be the simplest and safest solution. The only concern is it removes some public methods (see below), which were introduced in https://github.com/etcd-io/bbolt/pull/373, and has already included in v1.3.7. They were added not long ago (about 9 months ago), and it might be fine. We need to add a breaking change item in the changelog though.

https://github.com/etcd-io/bbolt/blob/233156a53aebc333337d8dc7b350271c9611591f/tx.go#L677-L794

**Comment by tjungblu** [comment]
> It is introduced in https://github.com/etcd-io/bbolt/commit/8ebdaf8d2ca98e26cc81d5ea33a1a86dd7cffc5f

@ahrtr we could also re-arrange the struct a bit for those things to be aligned again. It's certainly safer than to worry about which part of the stats need to take a lock or not.

**Comment by ahrtr** [comment]
> The variables that need to be aligned for this can be declared at the top of the struct, so it's not even a big change.

This is a good point. Refer to the second solution https://github.com/etcd-io/bbolt/pull/584

**Comment by tjungblu** [comment]
yep, good one and less risky :) Thanks!

**Comment by ahrtr** [comment]
thx @tjungblu 

Verified that https://github.com/etcd-io/bbolt/pull/584 fixed this issue using `qemu-user-static`.  @nekohasekai would you mind to verify the PR in your environment?

**Comment by ahrtr** [comment]
Resolved in [v1.3.8](https://github.com/etcd-io/bbolt/releases/tag/v1.3.8)
